### ISSUE TITLE
Consumer-only example

### ISSUE THREAD

**Issue by barryoneill** [ovotech/fs2-kafka#101]
Hi, I'm really interested in using this library (thank you for it!), but it appears to be designed around a consumer-producer mode only.  

If I wanted a composable stream - (i.e.) 

```scala
def subscribe(topic: String): Stream[F, MyModel] =
    for {
      ec <- consumerExecutionContextStream[F]
      record <- consumerStream[F]
                 .using(consumerSettings("myGroup", AutoOffsetReset.Latest)(ec))
                 .evalTap(_.subscribeTo(topic))
                 .flatMap(_.stream)
                 .mapAsync(maxConcurrent = 25) { event =>
                   ConcurrentEffect[F].pure(event.record)
                 }
    } yield MyModel(record.value())

```

This appears to work, but doesn't perform very well (Caveat, I'm an FP novice).  Could the lack of `commitBatchWithinF(...)` interfere with backpressure in this use case?   I can't see any calls that let me invoke commit()  periodically in the above example that don't break the composability of the function. 

In short - I'd love an example where there isn't any assumption about a producer. (even the scaladoc for the `commitBatchWithinF` hints it was designed around this use case), and I just can return the caller a `Stream[F, EventModel]`?

**Comment by vlovgr** [comment]
Thanks for raising this issue @barryoneill!

> it appears to be designed around a consumer-producer mode only

While the [quick example](https://ovotech.github.io/fs2-kafka/docs/quick-example) in the documentation uses both, consumers and producers can be used independently.

> Could the lack of `commitBatchWithinF(...)` interfere with backpressure in this use case?

It only adds additional backpressure, but you still have backpressure in the stream.

>  I can't see any calls that let me invoke commit() periodically in the above example that don't break the composability of the function.

When you `.flatMap(_.stream)` you end up with a `Stream[F, CommittableMessage[F]]`, where each `CommittableMessage[F]` contains the record offset and a way to commit it. These messages are typically batched together and committed together for performance reasons.

If you wanted to commit offsets, you'd have to return the `CommittableOffset[F]` part of the message, either as part of `MyModel`, or using a tuple or equivalent. You could do something like the following.

```scala
def subscribe(topic: String): Stream[F, (MyModel, CommittableOffset[F])] =
  consumerStream[F]
    .using(consumerSettings("myGroup", AutoOffsetReset.Latest))
    .evalTap(_.subscribeTo(topic))
    .flatMap(_.stream)
    .map { message =>
      val model = MyModel(message.record.value)
      val offset = message.committableOffset
      F.pure(model -> offset)
    }
```

If you then want to use that function and commit offsets, it's easiest to use one of the provided batch commit functions, for example `commitBatchWithin`. In the example below, we commit every 200 offsets, or 10 seconds, whichever happens first.

```scala
subscribe("topic")
  .map { case (_, offset) => offset }
  .through(commitBatchWithin(200, 10.seconds))
```

**Comment by barryoneill** [comment]
Hi - I appreciate the response.  I think the issue for me is the the very last piece, where I have to map to the offset in order to use one of the `commit*` methods, which makes my stream `Stream[F, Unit]`. 

In my case, this is a utility which is scanning a kafka topic for a particular event, and the Stream ends by returning information from the event.  By sinking using the `commit*` methods, I can't do that, I think?

**Comment by vlovgr** [comment]
> In my case, this is a utility which is scanning a kafka topic for a particular event, and the Stream ends by returning information from the event. By sinking using the `commit*` methods, I can't do that, I think?

Are you okay with committing offsets even if you haven't finished processing that particular event? If you commit the offset before processing finishes, you run the risk of crashing and not seeing that event again when you resume (i.e. you've no longer got an at-least-once guarantee).

If you're okay without the at-least-once delivery guarantee, and want to commit before processing of the record has finished, you need something like what was proposed in #79 (see [here](https://github.com/david-mcgillicuddy-ovo/fs2-kafka/blob/ea5fd12c5b007095f732bd94c3fe885f7e4e40d7/src/main/scala/fs2/kafka/package.scala#L150-L162) for the specific function, `commitBatchChunkWithPassthrough`).

**Comment by barryoneill** [comment]
Probably not for my use case.  I can refactor my application to move the processing I need within the `evalMap`, but I've learned lots through this little endeavor :)  

Thanks for your help!  I think we can close this.
